sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -263,6 +267,123 @@ case class GpuStringTrimRight(column: Expression, trimParameters: Option[Express
+  }
+
+  private def concatArrayCol( sep: Either[GpuScalar, GpuColumnVector],
+      cv: ColumnView): GpuColumnVector = {
+    sep match {
+      case Left(sepScalar) =>
+        withResource(GpuScalar.from("", StringType)) { emptyStrScalar =>
+          GpuColumnVector.from(cv.stringConcatenateListElements(sepScalar.getBase, emptyStrScalar,
+            false, false), dataType)
+        }
+      case Right(sepVec) =>
+        GpuColumnVector.from(cv.stringConcatenateListElements(sepVec.getBase), dataType)
+    }
+  }
+
+  private def processSingleColScalarSep(cv: ColumnVector): GpuColumnVector = {
+    // single column with scalar separator just replace any nulls with empty string
+    withResource(Scalar.fromString("")) { emptyStrScalar =>
+      GpuColumnVector.from(cv.replaceNulls(emptyStrScalar), dataType)
+    }
+  }
+
+  private def stringConcatSeparatorScalar(columns: ArrayBuffer[ColumnVector],
+      sep: GpuScalar): GpuColumnVector = {
+    withResource(Scalar.fromString("")) { emptyStrScalar =>
+      GpuColumnVector.from(ColumnVector.stringConcatenate(sep.getBase, emptyStrScalar,
+        columns.toArray[ColumnView], false), dataType)
+    }
+  }
+
+  private def stringConcatSeparatorVector(columns: ArrayBuffer[ColumnVector],
+      sep: GpuColumnVector): GpuColumnVector = {
+    // GpuOverrides doesn't allow only specifying a separator, you have to specify at
+    // least one value column
+    GpuColumnVector.from(ColumnVector.stringConcatenate(columns.toArray[ColumnView],
+      sep.getBase), dataType)
+  }
+
+  private def resolveColumnVectorAndConcatArrayCols(
+      expr: Expression,
+      numRows: Int,
+      sep: Either[GpuScalar, GpuColumnVector],
+      batch: ColumnarBatch): GpuColumnVector = {
+    withResourceIfAllowed(expr.columnarEval(batch)) {
+      case vector: GpuColumnVector =>
+        vector.dataType() match {
+          case ArrayType(st: StringType, nullable) => concatArrayCol(sep, vector.getBase)
+          case _ => vector.incRefCount()
+        }
+      case s: GpuScalar =>
+        s.dataType match {
+          case ArrayType(st: StringType, nullable) =>
+            // we have to first concatenate any array types
+            withResource(GpuColumnVector.from(s, numRows, s.dataType).getBase) { cv =>
+              concatArrayCol(sep, cv)
+            }
+          case _ => GpuColumnVector.from(s, numRows, s.dataType)
+        }
+      case other =>
+        throw new IllegalArgumentException(s"Cannot resolve a ColumnVector from the value:" +
+          s" $other. Please convert it to a GpuScalar or a GpuColumnVector before returning.")
+      }
+  }
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val numRows = batch.numRows()
+    var sep: Either[GpuScalar, GpuColumnVector] = null
+    try {
+      // get the separator and check for null scalar separator
+      val sepScalarNullOpt = withResourceIfAllowed(children.head.columnarEval(batch)) {
```

Comment:
Is there a point to using `withResourceIfAllowed` if every path within is going to increment the refcount?  My original suggestion to use it was so that we didn't need to close the scalar in the finally block, but since it's going out of it's way to do that anyway, it's wasteful to try to close something that will never actually close here.  We should either wrap the whole code block in the `withResourceIfAllowed` so the expr evaluation is always closed at the end, whether it's a scalar or vector, or remove it and the increfcounts.